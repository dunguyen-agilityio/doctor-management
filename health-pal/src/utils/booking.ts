import { BookingData } from '@/models/booking'

import { createDayjs } from './date'

export const formatBooking = ({ doctor, date, documentId, id, time, type }: BookingData) => {
  const { users_permissions_user, clinic, documentId: doctorDocId, id: doctorId } = doctor

  return {
    date: createDayjs(date),
    documentId,
    id,
    time,
    type,
    doctorName: users_permissions_user.name,
    doctorAvatar: users_permissions_user.avatar?.url,
    address: clinic.address,
    specialty: doctor.specialty.name,
    doctorId,
    doctorDocId,
  }
}
