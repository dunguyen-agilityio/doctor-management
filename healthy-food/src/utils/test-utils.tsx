import {
  RenderOptions,
  render,
  renderHook,
} from '@testing-library/react-native';

import { NavigationContainer } from '@react-navigation/native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const AllTheProviders = ({ children }: React.PropsWithChildren) => {
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });

  return (
    <QueryClientProvider client={queryClient}>
      <NavigationContainer>{children}</NavigationContainer>
    </QueryClientProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (renderCallback, options) =>
  renderHook(renderCallback, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render, customRenderHook as renderHook };
