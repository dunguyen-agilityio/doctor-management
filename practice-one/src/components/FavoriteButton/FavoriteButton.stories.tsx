import type { Meta, StoryObj } from '@storybook/react';

import { View } from 'react-native';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

import FavoriteButton from './index';

const queryClient = new QueryClient();

const meta = {
  title: 'FavoriteButton',
  component: FavoriteButton,
  argTypes: { onRefetch: { action: 'pressed' } },
  decorators: [
    (Story) => (
      <QueryClientProvider client={queryClient}>
        <View style={{ flex: 1 }}>
          <Story />
        </View>
      </QueryClientProvider>
    ),
  ],
} satisfies Meta<typeof FavoriteButton>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Favorite: Story = { args: { favorite: true, id: '1' } };

export const UnFavorite: Story = { args: { favorite: false, id: '1' } };
