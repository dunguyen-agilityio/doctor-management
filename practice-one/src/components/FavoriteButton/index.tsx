import { useState } from 'react';
import {
  ActivityIndicator,
  Platform,
  StyleSheet,
  ToastAndroid,
  TouchableOpacity,
} from 'react-native';

import { useMutation, useQueryClient } from '@tanstack/react-query';

import { COLOR } from '@/constants';

import { addFoodToFavorite, removeFoodToFavorite } from '@/services';

import Text from '../Text';

interface FavoriteButtonProps {
  favorite?: boolean;
  id: string;
  favoriteId?: string;
}

const FavoriteButton = ({ favorite, id, favoriteId }: FavoriteButtonProps) => {
  const queryClient = useQueryClient();
  const [hasFavorite, setHasFavorite] = useState(favorite);

  const { mutate, isPending } = useMutation({
    mutationFn: async () => {
      if (favorite && favoriteId) {
        await removeFoodToFavorite(favoriteId);
      } else {
        await addFoodToFavorite(id);
      }
    },
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ['foods-favorite'] });
      await queryClient.invalidateQueries({ queryKey: [`food-${id}`] });
      setHasFavorite((prev) => !prev);

      if (Platform.OS === 'android') {
        ToastAndroid.show('Updated success!', ToastAndroid.SHORT);
      }
    },
  });

  const handleFavorite = () => {
    mutate();
  };

  return (
    <TouchableOpacity
      onPress={handleFavorite}
      style={styles.button}
      disabled={isPending}
      testID="favorite-button"
    >
      {isPending ? (
        <ActivityIndicator
          size="small"
          color="#0000ff"
          testID="activity-indicator"
        />
      ) : (
        <Text variant="subtitle1" color={COLOR.WHITE}>
          {hasFavorite ? 'UnFavorites' : 'Add to Favorites'}
        </Text>
      )}
    </TouchableOpacity>
  );
};

export default FavoriteButton;

const styles = StyleSheet.create({
  button: {
    width: '100%',
    borderRadius: 9,
    paddingVertical: 9,
    marginTop: 27,
    backgroundColor: COLOR.PRIMARY,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
