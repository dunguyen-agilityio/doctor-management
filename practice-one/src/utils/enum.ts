export const isInEnum = <T extends Record<string, string | number>>(
  obj: T,
  key: unknown = null,
): key is keyof T => {
  if (key == null) return false;
  return Object.keys(obj).includes(String(key));
};

export const isArrayEqual = <T>(arr1: T[], arr2: T[]) =>
  arr1.length == arr2.length && arr1.every((item) => arr2.includes(item));
