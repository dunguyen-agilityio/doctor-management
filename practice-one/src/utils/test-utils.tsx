import { RenderOptions, render } from '@testing-library/react-native';

import { NavigationContainer } from '@react-navigation/native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const AllTheProviders = ({ children }: React.PropsWithChildren) => {
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });

  return (
    <QueryClientProvider client={queryClient}>
      <NavigationContainer>{children}</NavigationContainer>
    </QueryClientProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
