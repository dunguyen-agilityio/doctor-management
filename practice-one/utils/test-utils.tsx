import { ReactElement } from 'react';

import { NavigationContainer } from '@react-navigation/native';

import { RenderOptions, render } from '@testing-library/react-native';

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return <NavigationContainer>{children}</NavigationContainer>;
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
