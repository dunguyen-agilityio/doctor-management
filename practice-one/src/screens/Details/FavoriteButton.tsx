import { StyleSheet } from 'react-native';

import { Button, Text } from '@/components';

import { COLOR } from '@/constants';

import type { IFood } from '@/types';

import { useFavorite } from '@/hooks/useFavorite';

const FavoriteButton = ({ food }: { food: IFood }) => {
  const { id: foodId } = food;

  const { favorites, addToFavorite, removeFromFavorite } = useFavorite();

  const hasFavorite = favorites.some(({ id }) => foodId === id);

  const toggleFavorite = () => {
    hasFavorite ? removeFromFavorite(foodId) : addToFavorite(food);
  };

  return (
    <Button onPress={toggleFavorite} style={styles.button}>
      <Text variant="subtitle1" color={COLOR.WHITE}>
        {hasFavorite ? 'Unfavorite' : 'Add to Favorites'}
      </Text>
    </Button>
  );
};

export default FavoriteButton;

const styles = StyleSheet.create({
  button: {
    marginTop: 27,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
