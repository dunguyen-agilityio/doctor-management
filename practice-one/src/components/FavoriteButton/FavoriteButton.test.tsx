import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

import FavoriteButton from '@/components/FavoriteButton';

import { render, screen } from '@/utils/test-utils';

// Mock the updateFood service
jest.mock('@/services', () => ({
  updateFood: jest.fn(),
}));

describe('FavoriteButton Component', () => {
  let queryClient: QueryClient;

  beforeEach(() => {
    queryClient = new QueryClient();
  });

  const renderComponent = (favorite = false) => {
    return render(
      <QueryClientProvider client={queryClient}>
        <FavoriteButton favorite={favorite} id="123" />
      </QueryClientProvider>,
    );
  };

  it('renders the correct initial text', () => {
    renderComponent(false);
    expect(screen.getByText('Add to Favorites')).toBeTruthy();

    renderComponent(true);
    expect(screen.getByText('UnFavorites')).toBeTruthy();
  });

  it('calls the mutation function when clicked', async () => {});

  it('shows loading indicator when the mutation is pending', async () => {});

  it('updates the text after a successful mutation', async () => {});

  it('matches snapshot', () => {
    const tree = renderComponent(false).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
