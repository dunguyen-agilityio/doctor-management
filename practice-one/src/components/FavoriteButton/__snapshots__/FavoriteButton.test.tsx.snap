// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteButton Component matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#91C788",
      "borderRadius": 9,
      "justifyContent": "center",
      "marginTop": 27,
      "opacity": 1,
      "paddingVertical": 9,
      "width": "100%",
    }
  }
  testID="favorite-button"
>
  <Text
    style={
      [
        {
          "color": "#fff",
          "textTransform": undefined,
        },
        {
          "fontSize": 20,
          "fontWeight": "600",
          "lineHeight": 30,
        },
        undefined,
      ]
    }
  >
    Add to Favorites
  </Text>
</View>
`;
